pkg/controller/infra/infra_controller.go

```diff
@@ -118,99 +138,369 @@ type InfraController struct {
+	c.enqueueCluster(cluster)
+}
+
+func (c *InfraController) isInfraJob(job *v1batch.Job) bool {
+	controllerRef := metav1.GetControllerOf(job)
+	if controllerRef == nil {
+		return false
+	}
+	if controllerRef.Kind != clusterKind.Kind {
+		return false
+	}
+	return strings.HasPrefix(job.Name, jobPrefix)
+}
+
+func (c *InfraController) clusterForJob(job *v1batch.Job) (*clusteroperator.Cluster, error) {
+	controllerRef := metav1.GetControllerOf(job)
+	if controllerRef.Kind != clusterKind.Kind {
+		return nil, nil
+	}
+	cluster, err := c.clustersLister.Clusters(job.Namespace).Get(controllerRef.Name)
+	if err != nil {
+		return nil, err
+	}
+	if cluster.UID != controllerRef.UID {
+		// The controller we found with this Name is not the same one that the
+		// ControllerRef points to.
+		return nil, nil
+	}
+	return cluster, nil
+}
+
+func (c *InfraController) addJob(obj interface{}) {
+	job := obj.(*v1batch.Job)
+	if c.isInfraJob(job) {
+		c.logger.Debugf("enqueueing created infra job %s/%s", job.Namespace, job.Name)
+		c.enqueueJob(job)
+	}
+}
+
+func (c *InfraController) updateJob(old, obj interface{}) {
+	job := obj.(*v1batch.Job)
+	if c.isInfraJob(job) {
+		c.logger.Debugf("enqueueing updated infra job %s/%s", job.Namespace, job.Name)
+		c.enqueueJob(job)
+	}
+}
+
+func (c *InfraController) deleteJob(obj interface{}) {
+	job, ok := obj.(*v1batch.Job)
+	if !ok {
+		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("couldn't get object from tombstone %#v", obj))
+			return
+		}
+		job, ok = tombstone.Obj.(*v1batch.Job)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("tombstone contained object that is not a Job %#v", obj))
+			return
+		}
+	}
+	if !c.isInfraJob(job) {
+		return
+	}
+	cluster, err := c.clusterForJob(job)
+	if err != nil || cluster == nil {
+		utilruntime.HandleError(fmt.Errorf("could not get cluster for deleted job %s/%s", job.Namespace, job.Name))
+	}
 	c.enqueueCluster(cluster)
 }
 
 // Runs c; will not return until stopCh is closed. workers determines how many
 // clusters will be handled in parallel.
 func (c *InfraController) Run(workers int, stopCh <-chan struct{}) {
 	defer utilruntime.HandleCrash()
-	defer c.queue.ShutDown()
+	defer c.clusterQueue.ShutDown()
+	defer c.jobQueue.ShutDown()
 
-	c.logger.Infof("starting cluster controller")
-	defer c.logger.Infof("shutting down cluster controller")
+	c.logger.Infof("starting infra controller")
+	defer c.logger.Infof("shutting down infra controller")
 
-	if !controller.WaitForCacheSync("cluster", stopCh, c.clustersSynced) {
+	if !controller.WaitForCacheSync("infra", stopCh, c.clustersSynced, c.jobsSynced) {
+		c.logger.Errorf("Could not sync caches for infra controller")
 		return
 	}
 
 	for i := 0; i < workers; i++ {
-		go wait.Until(c.worker, time.Second, stopCh)
+		go wait.Until(c.clusterWorker, time.Second, stopCh)
+		go wait.Until(c.jobWorker, time.Second, stopCh)
```

Comment:
I am very cautious against having a separate queue for syncing the jobs. Why can't we just enqueue the cluster associated with the job when the controller notices an add/update/delete for the job? When syncing the cluster, the controller will look at the provisioning job for that cluster. Is the separate queue needed so that we can mark the probe time for the provisioning condition without having that cause another immediate sync of the cluster with a corresponding job check, causing a hot loop. If so, is there some other way to avoid the loop short of using a separate queue?

Reply:
We can use one queue. The way this PR started, I had no reference from the cluster to the actively running job, which meant that I couldn't just look at the cluster and immediately find the job. Now that we have a ref, I don't see a good reason to keep the 2 queues. Will change.